scrapy/utils/conf.py

```diff
@@ -1,36 +1,54 @@
 import os
 import sys
+import warnings
 from operator import itemgetter
 
 import six
 from six.moves.configparser import SafeConfigParser
 
+from scrapy.exceptions import ScrapyDeprecationWarning
+from scrapy.settings import BaseSettings
 from scrapy.utils.deprecate import update_classpath
 
 
-def build_component_list(base, custom, convert=update_classpath):
-    """Compose a component list based on a custom and base dict of components
-    (typically middlewares or extensions), unless custom is already a list, in
-    which case it's returned.
-    """
+def build_component_list(compdict, custom=None, convert=update_classpath):
+    """Compose a component list from a { class: order } dictionary."""
 
     def _check_components(complist):
         if len({convert(c) for c in complist}) != len(complist):
             raise ValueError('Some paths in {!r} convert to the same object, '
                              'please update your settings'.format(complist))
 
-    if isinstance(custom, (list, tuple)):
-        _check_components(custom)
-        return type(custom)(convert(c) for c in custom)
-
     def _map_keys(compdict):
         _check_components(compdict)
-        return {convert(k): v for k, v in six.iteritems(compdict)}
-
-    compdict = _map_keys(base)
-    compdict.update(_map_keys(custom))
-    items = (x for x in six.iteritems(compdict) if x[1] is not None)
-    return [x[0] for x in sorted(items, key=itemgetter(1))]
+        if isinstance(compdict, BaseSettings):
+            compbs = BaseSettings()
+            for k, v in six.iteritems(compdict):
+                prio = compdict.getpriority(k)
+                compbs.set(convert(k), v, priority=prio)
+            return compbs
```

Comment:
It took me a bit to understand this, I think this implementation is halfway between the old behavior and the new one, though I'm probably missing something.

If you call _check_components in a `compdict` that is a BaseSettings instance, you're making sure that no keys convert to the same path/value, regardless of their priority. In that case adding them converted into a new BaseSettings shouldn't make any difference since there aren't any clashes, we can stick to `{convert(k): v for k, v in six.iteritems(compdict)}` (there's no need to return a BaseSettings too).

Another possibility should be to not call _check_components in a BaseSettings, and check manually if there are any two keys that convert to the same path and also have the same priority so we can't tell which should be overridden first (commented in the description of #1267).

It's true that there weren't per-keys priorities before, users working with old paths should get an error if any two keys clash if they switch to scrapy 1.0, but it's probably a nicer deprecation for the old paths to users skipping 1.0 and updating directly to 1.1. If possible I would prefer the second option of checking if any keys that clash have the same priority or not thought I think it's not that critical, we can implement the first one and delete this `if` otherwise.